File: framework/client/servercomm.py
import requests
import shutil
import json


#class which describes the ways that the program can interact with the server.
class servercomm:
    #input is the name of the simulation (as known to the server)
    def __init__(self,simname):
        self.targetbase='http://192.168.2.14:5000/'
        self.simname=simname
        print("Server initialised for simulation '"+self.simname+"'.")
        self.started=False

    #tell server to start the simulation. Pass in the configuration file
    def StartSim(self,configfile):
        if not self.started:
            files= {'fileToUpload': open(configfile,'rb')}
            print(self.targetbase+'simulation/'+self.simname)
            postrequest = requests.post(self.targetbase+'simulation/'+self.simname, files=files)
            self.simid=postrequest.text
            self.base= self.targetbase +'simulation/'+ self.simname +'/'+ self.simid
            self.data_base=self.base+'/data/'
            self.started=True
            print("Simulation Started. ID="+self.simid)
        else:
            print("Error, simulation has already started")
            print("Simulation ID="+self.simid)

    #get the status of the simulation. Returns a dictionary containing 'status' (running,finished etc) and 'files' (list of results files)
    def GetStatus(self):
        if self.started:
            statusrequest = requests.get(self.base,stream=True)
            self.status=json.loads(statusrequest.text)
            return self.status
        else:
            print("Error: No simulation is running")

    #Downloads the file 'file_to_get' from the server and saves it to the temp file 'name_of_tmp_file'
    def GetDataFile(self,file_to_get,name_of_tmp_file):
        if self.started:
            filerequest=requests.get(self.data_base + file_to_get,stream=True)
            with open(name_of_tmp_file,'wb') as f:
                for chunk in filerequest.iter_content(1024):
                    f.write(chunk)
        else:
            print("Error: No simulation is running")

    # delete the file 'file' from the server
    def DeleteFile(self,file):
        if self.started:
            deletefilerequest=requests.delete(self.data_base+file)
            print("Deleted file: '"+file+"'")
        else:
            print("Error: No simulation is running")

    #delete all the simulation's output files from the server
    def DeleteSim(self):
        if self.started:
            deletefilerequest=requests.delete(self.base)
            print("Deleted Simulation")
            self.started=False
        else:
            print("Error: No simulation to be deleted")

    #returns whether the server class object has started a simulation
    def IsStarted(self):
        return self.started

File: framework/web/src/server-comm.js
import axios from "axios";

class ServerComm {
    constructor(simName, serverAddress) {
        // this.targetBase = serverAddress || 'http://192.168.2.14:5000/'; // Wee Archie 1
        this.targetBase = serverAddress || 'http://192.168.2.25:5000/'; // Wee Archie 2

        this.simName = simName;

        console.info(`Server address is ${this.targetBase}.`);
        console.info(`Server initialised for simulation ${this.simName}.`);
        this.started = false;
        this.cancelTokenSource = axios.CancelToken.source();
        this.cancelToken  = this.cancelTokenSource.token;
    }

    // configFile must be a readable Stream
    async startSim(configFile)
    {
        if(!this.started) {
            // Alternative to locally provided data configs?
            const simAddress = `${this.targetBase}simulation/${this.simName}`
            console.info(`Posting to ${simAddress}`);

            const formData = new FormData();
            formData.append('fileToUpload', new Blob([configFile]));
            return await axios.post(
                simAddress,
                formData,
                {cancelToken: this.cancelToken}
            ).then( body => {
                this.simId = body;
                this.base = `${this.targetBase}simulation/${this.simName}/${this.simId}`;
                this.dataBase = `${this.base}/data/`;
                this.started = true;
                console.info(`Simulation started with ID = ${this.simId}`);
            })
            .catch(err => { console.error(`Simulation ${this.simId} failed
                with error ${err}`) });
        } else {
            console.error(`Simulation has already started.
                Existing Simulation ID: ${this.simId}`);
        }
    }

    async getStatus() {
        if (this.started) {
            return await axios.get(this.base, {cancelToken: this.cancelToken})
            .then( body => JSON.parse(body) )
            .catch(err => {console.error(`Request failed for ${this.simId}
                with error ${err}`)});
        } else {
            console.error(`No Simulation running`);
        }
    }

    async getDataFile(fileName) {
        if(this.started) {
            return await axios.get(this.dataBase + fileName,
            {cancelToken: this.cancelToken})
            .catch( err => { console.error(`Request failed getting \
                ${this.simId}/${fileName} with error
                ${err}`) });
        } else {
            console.error(`No Simulation running`);
        }
    }

    async deleteDataFile(fileName) {
        if(this.started) {
            return await axios.delete(this.dataBase + fileName,
            {cancelToken: this.cancelToken})
            .then( () => { console.info(`Deleted ${this.simId}/${fileName}`) })
            .catch( err => { console.error(`Request failed deleting \
                ${this.simId}/${fileName} with error
                ${err}`) });
        } else {
            console.error(`No Simulation running`);
        }
    }

    async deleteSim() {
        if(this.started) {
            return await axios.delete(this.base,
            {cancelToken: this.cancelToken})
            .then( () => { console.info(`Deleted Simulation ${this.simId}`) })
            .catch( err => { console.error(`Request failed deleting \
                ${this.simId} with error
                ${err}`) });
        } else {
            console.error(`No Simulation running`);
        }
    }

    isStarted() {
        return this.started;
    }

    cancel() {
        this.cancelTokenSource.cancel();
    }
}

export { ServerComm as default};
